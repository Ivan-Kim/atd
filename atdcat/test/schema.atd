(* Input ATD file for testing the translation to JSON Schema *)

type different_kinds_of_things = [
  | Root
  | Thing of int
  | WOW <json name="wow">
  | Amaze <json name="!!!"> of string list
]

type root = {
  id <json name="ID">: string;
  items: int list list;
  ?maybe: int option;
  ~extras: int list;
  ~answer: int;
  aliased: alias;
  point: (float * float);
  kinds: different_kinds_of_things list;
  assoc1: (float * int) list;
  assoc2: (string * int) list <json repr="object">;
  ~options: int option list;
  ~nullables: int nullable list;
}

type alias = int list

type pair = (string * int)
