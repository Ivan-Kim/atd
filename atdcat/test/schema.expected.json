{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "description": "Translated by atdcat from 'schema.atd'",
  "type": "object",
  "required": [
    "ID", "items", "aliased", "point", "kinds", "assoc1", "assoc2"
  ],
  "properties": {
    "ID": { "type": "string" },
    "items": {
      "type": "array",
      "items": { "type": "array", "items": { "type": "integer" } }
    },
    "maybe": { "type": "integer" },
    "extras": { "type": "array", "items": { "type": "integer" } },
    "answer": { "type": "integer" },
    "aliased": { "$ref": "#/definitions/alias" },
    "point": {
      "type": "array",
      "minItems": 2,
      "additionalItems": false,
      "items": [ { "type": "number" }, { "type": "number" } ]
    },
    "kinds": {
      "type": "array",
      "items": { "$ref": "#/definitions/different_kinds_of_things" }
    },
    "assoc1": {
      "type": "array",
      "items": {
        "type": "array",
        "minItems": 2,
        "additionalItems": false,
        "items": [ { "type": "number" }, { "type": "integer" } ]
      }
    },
    "assoc2": {
      "type": "array",
      "items": {
        "type": "array",
        "minItems": 2,
        "additionalItems": false,
        "items": [ { "type": "string" }, { "type": "integer" } ]
      }
    },
    "options": {
      "type": "array",
      "items": {
        "oneOf": [
          {
            "type": "array",
            "minItems": 2,
            "additionalItems": false,
            "items": [ { "const": "Some" }, { "type": "integer" } ]
          },
          { "const": "None" }
        ]
      }
    },
    "nullables": {
      "type": "array",
      "items": { "type": [ "integer", "null" ] }
    }
  },
  "definitions": {
    "different_kinds_of_things": {
      "oneOf": [
        { "const": "Root" },
        {
          "type": "array",
          "minItems": 2,
          "additionalItems": false,
          "items": [ { "const": "Thing" }, { "type": "integer" } ]
        },
        { "const": "wow" },
        {
          "type": "array",
          "minItems": 2,
          "additionalItems": false,
          "items": [
            { "const": "!!!" },
            { "type": "array", "items": { "type": "string" } }
          ]
        }
      ]
    },
    "alias": { "type": "array", "items": { "type": "integer" } },
    "pair": {
      "type": "array",
      "minItems": 2,
      "additionalItems": false,
      "items": [ { "type": "string" }, { "type": "integer" } ]
    }
  }
}
